crates/holochain/src/conductor/conductor/share.rs

```diff
@@ -0,0 +1,83 @@
+use std::sync::Arc;
+
+/// A clonable thread safe read write lock designed to make it hard to create dead locks
+/// or hold long long lived locks.
+pub struct RwShare<T>(Arc<parking_lot::RwLock<T>>);
+
+impl<T> Clone for RwShare<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<T> RwShare<T> {
+    /// Create a new shareable lock
+    pub fn new(value: T) -> Self {
+        Self(Arc::new(parking_lot::RwLock::new(value)))
+    }
+
+    /// Get a shared reference to the value. This will not block other readers
+    /// but will block writers.
+    /// This should never be used recursively or held over awaits
+    /// or held for a long time.
+    pub fn share_ref<R, F>(&self, f: F) -> R
+    where
+        F: FnOnce(&T) -> R,
+    {
+        let t = self
+            .0
+            // First we try to get a fair reader that won't starve writers.
+            .try_read_for(std::time::Duration::from_millis(100))
+            // The lock is taking a little longer then we'd like so print an info
+            // and try for a further 30 seconds.
+            .or_else(|| {
+                tracing::info!(
+                    "Took over 100ms to get a RwShare reader. Conductor might be over utilized"
+                );
+                self.0.try_read_for(std::time::Duration::from_secs(30))
+            })
+            // However if that fails we may be in a recursive reader dead lock so we will try for
+            // a recursive reader that may starve writers.
+            .or_else(|| {
+                tracing::warn!("Failed to get fair reader, trying for recursive reader");
+                self.0
+                    .try_read_recursive_for(std::time::Duration::from_secs(60))
+            })
+            // Now we are probably at a deadlock or a really long held lock so print an error.
+            .unwrap_or_else(|| {
+                tracing::error!(
+                    "Failed to get a RwShare read lock for over 120s this could be a dead lock"
+                );
+                self.0.read_recursive()
+            });
```

Comment:
This is philosophically counter to fail-fast. Do we want to allow recursive read locks? If holochain hobbles along and sort-of works after first waiting for 30 seconds, it feels like that just results in debugging onus.

I like the warning message after 100 ms, but maybe it should just panic after the 30 seconds?

Reply:
I worry that this could happen when someone runs holochain on a slow device or with other applications running. But I could add a panic after 5 mins. I don't think it should ever be locked for that long.

Reply:
Feels pretty long to me, but I'm fine setting it like that to start, and we can iterate as needed : )